### PR TITLE
Restore usage of open API safely

### PR DESCRIPTION
**Pull request by weizman**
A need for `open` usage was surfaced by Snow's first ever user MetaMask (https://github.com/MetaMask/metamask-extension/pull/16722) and therefore disabling it completely can no longer be an option.

`open` API leaves Snow very vulnerable to same origin attacks.

This PR comes to restore usage of `open` API while defending Snow from attacks against it.

This PR takes into consideration all possible attacks that are known to us via `open` API ([1](https://github.com/LavaMoat/snow/issues/2#issue-1252782135), [2](https://github.com/LavaMoat/snow/issues/2#issuecomment-1332422204) and [3](https://github.com/LavaMoat/snow/issues/2#issuecomment-1333436187) of #2)

Here's a full explanation of the solution [#2-1337593790](https://github.com/LavaMoat/snow/issues/2#issuecomment-1337593790)

NOTE: don't be alarmed by the amount of changed files, all files under `test/` folder (except for `open.js`) had a recurring line minorly modified to better suit new needs in this PR, when disregarding those you're left with just a few files to review.

### CODE REVIEW

File: src/open.js [hunk 1]
```diff
@@ -1,20 +1,84 @@
-const {slice, Function} = require('./natives');
-const {warn, WARN_OPEN_API_DISABLED} = require('./log');
+const {stringToLowerCase, stringStartsWith, slice, Function, Object, Reflect, Proxy, Map} = require('./natives');
+const {warn, WARN_OPEN_API_LIMITED, WARN_OPEN_API_URL_ARG_JAVASCRIPT_SCHEME} = require('./log');
 
-function hookOpen(win, cb) {
-    const realOpen = win.open;
-    win.open = function() {
+const openeds = new Map();
```
Comment (naugtur) [inline src/open.js:4]: a more descriptive name here would have caused one less WTF on line 11 ;)

File: src/open.js [hunk 2]
```diff
@@ -1,20 +1,84 @@
...
+
+function hookMessageEvent(win) {
+    const desc = Object.getOwnPropertyDescriptor(win.MessageEvent.prototype, 'source');
+    const get = desc.get;
+    desc.get = function() {
+        const source = get.call(this);
+        return openeds.get(source) || source;
+    };
+    Object.defineProperty(win.MessageEvent.prototype, 'source', desc);
```
Comment (naugtur) [inline src/open.js:13]: I'm not sure why this works yet, but assuming it does, great stuff!

File: src/open.js [hunk 3]
```diff
@@ -1,20 +1,84 @@
...
+            }
         }
 
-        const opened = Function.prototype.apply.call(realOpen, this, args);
+        const opened = Function.prototype.call.call(native, this, url, target, windowFeatures);
         cb(opened);
-        return opened;
-    }
+        const p = proxy(win, opened);
+        openeds.set(opened, p);
```
Comment (naugtur) [inline src/open.js:74]: It's possible openeds could be a WeakMap, looking at how it's used.

File: src/open.js [hunk 4]
```diff
@@ -1,20 +1,84 @@
...
+    const get = desc.get;
+    desc.get = function() {
+        const source = get.call(this);
+        return openeds.get(source) || source;
+    };
+    Object.defineProperty(win.MessageEvent.prototype, 'source', desc);
+}
+
+function proxy(win, opened) {
+    const target = {};
```
[the remark this answers is not in the document]
Comment (weizman) [reply src/open.js:17]: In case what?

File: src/open.js (hunk 1 above)
Comment (weizman) [reply src/open.js:4]: one look at line 74 makes it very clear I believe :) Not the best name though, I agree

File: src/open.js (hunk 3 above)
Comment (weizman) [reply src/open.js:78]: `win` refers to `WindowProxy` objects that are being referenced within scopes formed by Snow and therefore will never be garbage collected AFAIU so `WeakMap` special trait won't be applied here and will effectively act similarly to how `Map` does.

**Pull request merged**